File: python/src/com/sweetmanor/algorithm/graph/weighted_graph.py
# 《算法精粹》
# 加权图

from dataclasses import dataclass
from queue import PriorityQueue
from typing import TypeVar, Generic, List, Tuple, Optional, Dict

from graph import Graph
from weighted_edge import WeightedEdge

V = TypeVar('V')
WeightedPath = List[WeightedEdge]


@dataclass
class DijkstraNode:
    vertex: int
    distance: float

    def __lt__(self, other: 'DijkstraNode') -> bool:
        return self.distance < other.distance

    def __eq__(self, other: 'DijkstraNode') -> bool:
        return self.distance == other.distance


class WeightedGraph(Generic[V], Graph[V]):

    def __init__(self, vertices: List[V] = []) -> None:
        self._vertices = vertices
        self._edges: List[List[WeightedEdge]] = [[] for _ in vertices]

    def add_edge_by_indices(self, u: int, v: int, weight: float) -> None:
        edge = WeightedEdge(u, v, weight)
        self.add_edge(edge)

    def add_edge_by_vertices(self, first: V, second: V, weight: float) -> None:
        u = self._vertices.index(first)
        v = self._vertices.index(second)
        self.add_edge_by_indices(u, v, weight)

    def neighbors_for_index_with_weights(self, index: int) -> List[Tuple[V, float]]:
        distance_tuples: List[Tuple[V, float]] = []
        for edge in self.edges_for_index(index):
            distance_tuples.append((self.vertex_at(edge.v), edge.weight))
        return distance_tuples

    def __str__(self):
        desc: str = ""
        for i in range(self.vertex_count):
            desc += f"{self.vertex_at(i)} -> {self.neighbors_for_index_with_weights(i)}\n"
        return desc


def dijkstra(wg: WeightedGraph[V], root: V) -> Tuple[List[Optional[float]], Dict[int, WeightedEdge]]:
    """ Dijkstra 算法 """

    first = wg.index_of(root)  # 获取索引

    # 距离数组，起点为0
    distances: List[Optional[float]] = [None] * wg.vertex_count
    distances[first] = 0

    # 路径映射
    path_dict: Dict[int, WeightedEdge] = {}

    # 优先队列存放待遍历顶点
    pq: PriorityQueue[DijkstraNode] = PriorityQueue()
    pq.put(DijkstraNode(first, 0))

    while not pq.empty():
        u = pq.get().vertex
        dist_u = distances[u]
        # 遍历邻接边
        for we in wg.edges_for_index(u):
            dist_v = distances[we.v]  # 邻接边距离
            # 邻接边未访问或发现更短距离
            if dist_v is None or dist_v > we.weight + dist_u:
                distances[we.v] = we.weight + dist_u  # 更新距离
                path_dict[we.v] = we
                pq.put(DijkstraNode(we.v, we.weight + dist_u))

    return distances, path_dict


def distance_array_to_vertex_dict(wg: WeightedGraph[V], distances: List[Optional[float]]) -> Dict[V, Optional[float]]:
    distance_dict: Dict[V, Optional[float]] = {}
    for i in range(len(distances)):
        distance_dict[wg.vertex_at(i)] = distances[i]
    return distance_dict


def path_dict_to_path(start: int, end: int, path_dict: Dict[int, WeightedEdge]) -> WeightedPath:
    if len(path_dict) == 0:
        return []

    # 逆序添加路径
    edge_path: WeightedPath = []
    e = path_dict[end]
    edge_path.append(e)
    while e.u != start:
        e = path_dict[e.u]
        edge_path.append(e)

    return list(reversed(edge_path))


def mst(wg: WeightedGraph[V], start: int = 0) -> Optional[WeightedPath]:
    """ Prim算法：属于贪心算法 """
    # 校验入参，超出顶点列表范围时返回 None
    if start > wg.vertex_count - 1 or start < 0:
        return None

    path: WeightedPath = []  # 最终路径

    # 使用优先队列查找最小权重的边，使用列表记录已加入连通图顶点（等同于图遍历中的Set）
    pq: PriorityQueue[WeightedEdge] = PriorityQueue()
    visited = [False] * wg.vertex_count

    # 定义内部函数，方便后面多处引用
    def visit(index: int):
        visited[index] = True  # 标记顶点已访问
        for edge in wg.edges_for_index(index):
            # 遍历邻近边，未访问的加入优先队列
            if not visited[edge.v]:
                pq.put(edge)

    visit(start)  # 处理起始顶点
    # 优先队列不为空时继续处理
    while not pq.empty():
        edge = pq.get()
        # 对未访问的邻近边进行处理
        if not visited[edge.v]:
            path.append(edge)
            visit(edge.v)

    return path


def total_weight(wp: WeightedPath) -> float:
    return sum([e.weight for e in wp])


def print_weighted_path(wg: WeightedGraph, wp: WeightedPath) -> None:
    for edge in wp:
        print(f"{wg.vertex_at(edge.u)} {edge.weight} -> {wg.vertex_at(edge.v)}")
    print(f"总计权重：{total_weight(wp)}")


if __name__ == "__main__":
    city_graph = WeightedGraph(["Seattle", "San Francisco", "Los Angeles", "Riverside", "Phoenix",
                                "Chicago", "Boston", "New York", "Atlanta", "Miami",
                                "Dallas", "Houston", "Detroit", "Philadelphia", "Washington"])

    city_graph.add_edge_by_vertices("Seattle", "Chicago", 1737)
    city_graph.add_edge_by_vertices("Seattle", "San Francisco", 678)
    city_graph.add_edge_by_vertices("San Francisco", "Riverside", 386)
    city_graph.add_edge_by_vertices("San Francisco", "Los Angeles", 348)
    city_graph.add_edge_by_vertices("Los Angeles", "Riverside", 50)

    city_graph.add_edge_by_vertices("Los Angeles", "Phoenix", 357)
    city_graph.add_edge_by_vertices("Riverside", "Phoenix", 307)
    city_graph.add_edge_by_vertices("Riverside", "Chicago", 1704)
    city_graph.add_edge_by_vertices("Phoenix", "Dallas", 887)
    city_graph.add_edge_by_vertices("Phoenix", "Houston", 1015)

    city_graph.add_edge_by_vertices("Dallas", "Chicago", 805)
    city_graph.add_edge_by_vertices("Dallas", "Atlanta", 721)
    city_graph.add_edge_by_vertices("Dallas", "Houston", 225)
    city_graph.add_edge_by_vertices("Houston", "Atlanta", 702)
    city_graph.add_edge_by_vertices("Houston", "Miami", 968)

    city_graph.add_edge_by_vertices("Atlanta", "Chicago", 588)
    city_graph.add_edge_by_vertices("Atlanta", "Washington", 543)
    city_graph.add_edge_by_vertices("Atlanta", "Miami", 604)
    city_graph.add_edge_by_vertices("Miami", "Washington", 923)
    city_graph.add_edge_by_vertices("Chicago", "Detroit", 238)

    city_graph.add_edge_by_vertices("Detroit", "Boston", 613)
    city_graph.add_edge_by_vertices("Detroit", "Washington", 396)
    city_graph.add_edge_by_vertices("Detroit", "New York", 482)
    city_graph.add_edge_by_vertices("Boston", "New York", 190)
    city_graph.add_edge_by_vertices("New York", "Philadelphia", 81)

    city_graph.add_edge_by_vertices("Philadelphia", "Washington", 123)

    print(city_graph)

    path: Optional[WeightedPath] = mst(city_graph)
    if path is None:
        print("No solution found!")
    else:
        print_weighted_path(city_graph, path)

    print()

    distances, path_dict = dijkstra(city_graph, "Los Angeles")
    name_distance: Dict[str, Optional[int]] = distance_array_to_vertex_dict(city_graph, distances)
    print("Distances from Los Angeles:")
    for key, value in name_distance.items():
        print(f"{key} : {value}")

    print()

    print("Shortest path from Los Angeles to Boston:")
    path = path_dict_to_path(city_graph.index_of("Los Angeles"), city_graph.index_of("Boston"), path_dict)
    print_weighted_path(city_graph, path)

File: java/src/main/java/com/sweetmanor/algorithm/graph/WeightedGraph.java
package com.sweetmanor.algorithm.graph;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.*;
import java.util.function.IntConsumer;

/**
 * 加权图 —— 《算法精粹》
 *
 * @author ijlhjj
 * @version 1.0 2023-12-19
 */
public class WeightedGraph<V> extends Graph<V, WeightedEdge> {
    private static final Logger logger = LogManager.getLogger();

    public WeightedGraph(List<V> vertices) {
        super(vertices);
    }

    public void addEdge(WeightedEdge e) {
        edges.get(e.u).add(e);
        edges.get(e.v).add(e.reversed());
    }

    public void addEdge(int u, int v, double weight) {
        addEdge(new WeightedEdge(u, v, weight));
    }

    public void addEdge(V u, V v, double weight) {
        addEdge(indexOf(u), indexOf(v), weight);
    }

    public double totalWeight(List<WeightedEdge> path) {
        return path.stream().mapToDouble(we -> we.weight).sum();
    }

    public void printWeightedPath(List<WeightedEdge> path) {
        for (WeightedEdge e : path)
            logger.info("{} {} -> {}", vertexAt(e.u), e.weight, vertexAt(e.v));
        logger.info("总计权重：{}", totalWeight(path));
    }

    /**
     * Prim算法：属于贪心算法
     */
    public List<WeightedEdge> mst(int start) {
        List<WeightedEdge> path = new LinkedList<>(); //最终路径

        //校验入参，超出顶点列表范围时返回空路径
        if (start < 0 || start > getVertexCount() - 1)
            return path;

        //使用优先队列查找最小权重的边，使用数组记录已加入连通图顶点（等同于图遍历中的Set）
        PriorityQueue<WeightedEdge> priorityQueue = new PriorityQueue<>();
        boolean[] visited = new boolean[getVertexCount()];

        //定义内部函数，方便后面多处引用
        IntConsumer visit = index -> {
            visited[index] = true; //标记顶点已访问
            for (WeightedEdge edge : edgesOf(index)) {
                //遍历邻近边，未访问的加入优先队列
                if (!visited[edge.v])
                    priorityQueue.offer(edge);
            }
        };

        visit.accept(start); //处理起始顶点
        //优先队列不为空时继续处理
        while (!priorityQueue.isEmpty()) {
            WeightedEdge edge = priorityQueue.poll();
            //对未访问的邻近边进行处理
            if (!visited[edge.v]) {
                path.add(edge);
                visit.accept(edge.v);
            }
        }

        return path;
    }

    /**
     * Dijkstra 算法
     */
    public DijkstraResult dijkstra(V root) {
        int first = indexOf(root); //获取索引

        int vertexCount = getVertexCount();

        //距离数组，起点为0
        double[] distances = new double[vertexCount];
        distances[first] = 0;
        //已访问标记
        boolean[] visited = new boolean[vertexCount];
        visited[first] = true;

        //路径映射
        Map<Integer, WeightedEdge> pathMap = new HashMap<>();
        //优先队列存放待遍历顶点
        PriorityQueue<DijkstraNode> priorityQueue = new PriorityQueue<>();
        priorityQueue.offer(new DijkstraNode(first, 0));

        while (!priorityQueue.isEmpty()) {
            int u = priorityQueue.poll().vertex;
            double distU = distances[u];
            //遍历邻接边
            for (WeightedEdge e : edgesOf(u)) {
                double pathWeight = e.weight + distU; //邻接边距离
                double distV = distances[e.v];
                //邻接边未访问或发现更短距离
                if (!visited[e.v] || distV > pathWeight) {
                    visited[e.v] = true;
                    distances[e.v] = pathWeight; //更新距离
                    pathMap.put(e.v, e);
                    priorityQueue.offer(new DijkstraNode(e.v, pathWeight));
                }
            }
        }

        return new DijkstraResult(distances, pathMap);
    }

    public Map<V, Double> distanceArrayToDistanceMap(double[] distances) {
        Map<V, Double> distanceMap = new HashMap<>();
        for (int i = 0; i < distances.length; i++) {
            distanceMap.put(vertexAt(i), distances[i]);
        }
        return distanceMap;
    }

    public static List<WeightedEdge> pathMapToPath(int start, int end, Map<Integer, WeightedEdge> pathMap) {
        if (pathMap.isEmpty())
            return List.of();

        //逆序添加路径
        List<WeightedEdge> path = new LinkedList<>();
        WeightedEdge edge = pathMap.get(end);
        path.add(edge);

        while (edge.u != start) {
            edge = pathMap.get(edge.u);
            path.add(edge);
        }

        Collections.reverse(path);

        return path;
    }

    public static void main(String[] args) {
        WeightedGraph<String> cityGraph = new WeightedGraph<>(List.of("Seattle", "San Francisco", "Los Angeles", "Riverside", "Phoenix", "Chicago", "Boston", "New York", "Atlanta", "Miami", "Dallas", "Houston", "Detroit", "Philadelphia", "Washington"));

        cityGraph.addEdge("Seattle", "Chicago", 1737);
        cityGraph.addEdge("Seattle", "San Francisco", 678);
        cityGraph.addEdge("San Francisco", "Riverside", 386);
        cityGraph.addEdge("San Francisco", "Los Angeles", 348);
        cityGraph.addEdge("Los Angeles", "Riverside", 50);

        cityGraph.addEdge("Los Angeles", "Phoenix", 357);
        cityGraph.addEdge("Riverside", "Phoenix", 307);
        cityGraph.addEdge("Riverside", "Chicago", 1704);
        cityGraph.addEdge("Phoenix", "Dallas", 887);
        cityGraph.addEdge("Phoenix", "Houston", 1015);

        cityGraph.addEdge("Dallas", "Chicago", 805);
        cityGraph.addEdge("Dallas", "Atlanta", 721);
        cityGraph.addEdge("Dallas", "Houston", 225);
        cityGraph.addEdge("Houston", "Atlanta", 702);
        cityGraph.addEdge("Houston", "Miami", 968);

        cityGraph.addEdge("Atlanta", "Chicago", 588);
        cityGraph.addEdge("Atlanta", "Washington", 543);
        cityGraph.addEdge("Atlanta", "Miami", 604);
        cityGraph.addEdge("Miami", "Washington", 923);
        cityGraph.addEdge("Chicago", "Detroit", 238);

        cityGraph.addEdge("Detroit", "Boston", 613);
        cityGraph.addEdge("Detroit", "Washington", 396);
        cityGraph.addEdge("Detroit", "New York", 482);
        cityGraph.addEdge("Boston", "New York", 190);
        cityGraph.addEdge("New York", "Philadelphia", 81);

        cityGraph.addEdge("Philadelphia", "Washington", 123);

        logger.info(cityGraph);

        List<WeightedEdge> mst = cityGraph.mst(0);
        cityGraph.printWeightedPath(mst);

        logger.info("\n");

        DijkstraResult dijkstraResult = cityGraph.dijkstra("Los Angeles");
        Map<String, Double> nameDistance = cityGraph.distanceArrayToDistanceMap(dijkstraResult.distances);
        logger.info("Distances from Los Angeles:");
        nameDistance.forEach((name, distance) -> logger.info("{} : {}", name, distance));

        logger.info("\n");

        logger.info("Shortest path from Los Angeles to Boston:");
        List<WeightedEdge> path = pathMapToPath(cityGraph.indexOf("Los Angeles"), cityGraph.indexOf("Boston"), dijkstraResult.pathMap);
        cityGraph.printWeightedPath(path);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < getVertexCount(); i++) {
            sb.append(vertexAt(i));
            sb.append(" -> ");
            sb.append(Arrays.toString(edgesOf(i).stream().map(we -> "(" + vertexAt(we.v) + ", " + we.weight + ")").toArray()));
            sb.append(System.lineSeparator());
        }
        return sb.toString();
    }

    public record DijkstraNode(int vertex, double distance) implements Comparable<DijkstraNode> {
        @Override
        public int compareTo(DijkstraNode other) {
            Double mine = distance;
            Double theirs = other.distance;
            return mine.compareTo(theirs);
        }
    }

    public record DijkstraResult(double[] distances, Map<Integer, WeightedEdge> pathMap) {
    }

}
